Add evictor based loadbalancer caching

OlgaMaciaszek opened the pull request:
Provides an Evictor-based default Caching mechanism for Spring Cloud LoadBalancer.

ryanjbaxter commented on spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/config/LoadBalancerCacheAutoConfiguration.java:
```diff
@@ -77,24 +80,38 @@ LoadBalancerCaffeineWarnLogger caffeineWarnLogger() {
+	@Configuration(proxyBeanMethods = false)
+	@ConditionalOnMissingClass("com.github.benmanes.caffeine.cache.Caffeine")
+	@ConditionalOnClass(ConcurrentMapWithTimedEviction.class)
+	protected static class EvictorLoadBalancerCacheManagerConfiguration {
```
Don't you need Evictor on the classpath?

ryanjbaxter commented on docs/src/main/asciidoc/spring-cloud-commons.adoc:
```diff
@@ -803,19 +803,26 @@ in the following sections:
-via `DiscoveryClient` each time it has to choose an instance, we provide a https://github.com/ben-manes/caffeine[Caffeine-backed]
-implementation.
-To make use of it, you need to have `com.github.ben-manes.caffeine:caffeine` in the classpath.
+Apart from the basic `ServiceInstanceListSupplier` implementation that retrieves instances via `DiscoveryClient` each time it has to choose an instance, we provide two caching implementations:
```
I think we need to organize this caching documentation a little better.

Can we break things into 3 sections

1. No caching library on the classpath
2. Caffeine on the classpath
3. Evictor on the classpath

ryanjbaxter commented on docs/src/main/asciidoc/spring-cloud-commons.adoc:
```diff
@@ -803,19 +803,26 @@ in the following sections:
 The default setup includes `expireAfterWrite` set to 30 seconds and records set to soft references.
+The default `initialCapacity` is `256`.
+
+If you do not have Caffeine in the classpath, but have `com.stoyanr:evictor` in the classpath, Evictor-based implementation will be used.
```
I wouldn't say if Evictor is on the classpath because it will be

OlgaMaciaszek replied:
yes, but if someone is using the dependency directly, they will not have it. Will try to write it more expressely.

OlgaMaciaszek commented:
> Why all the flatten-pom changes?

The flattened-poms should not be in master at all. I was getting diffs on them as they were, so I removed them and added to `.gitignore`.

ryanjbaxter commented on docs/src/main/asciidoc/spring-cloud-commons.adoc:
```diff
@@ -803,25 +803,35 @@ in the following sections:
+
+==== Default LoadBalancer Cache Implementation
+
+If you do not have Caffeine in the classpath, but have https://github.com/stoyanr[`com.stoyanr:evictor`] (which comes automatically with `spring-cloud-starter-loadbalancer`), the default, Evictor-based, implementation will be used.
```
> If you do not have Caffeine in the classpath, but have https://github.com/stoyanr[`com.stoyanr:evictor`] (which comes automatically with `spring-cloud-starter-loadbalancer`)

I would leave all this stuff about Evictor out.  Just tell them how to 

1.  Configure the caching parameters
2. Use Caffeine instead of our own cache implementation

ryanjbaxter commented on docs/src/main/asciidoc/spring-cloud-commons.adoc:
```diff
@@ -803,25 +803,35 @@ in the following sections:
 
 WARNING: Although the basic, non-cached, implementation is useful for prototyping and testing, it's much less efficient
 than the cached versions, so we recommend always using the cached version in production.
+If you are using `spring-cloud-starter-loadbalancer`, the cache will be set up and enabled for you by default, but if you are using just the loadbalancer module dependency (and not the starter), that will not be done.
```
They should always be using the starter so you can remove this

ryanjbaxter commented on spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/cache/DefaultCache.java:
```diff
@@ -0,0 +1,175 @@
+ * @see ConcurrentMapWithTimedEviction
+ * @see ConcurrentMapCache
+ */
+public class DefaultCache extends AbstractValueAdaptingCache {
```
I would call this something like `DefaultLoadBalancerCache`
